test/test_common/real_threads_test.cc

```diff
@@ -0,0 +1,98 @@
+#include "real_threads_test.h"
+
+#include "absl/synchronization/barrier.h"
+#include "utility.h"
+
+namespace Envoy {
+namespace Thread {
+
+RealThreadsTestBase::RealThreadsTestBase(uint32_t num_threads)
+    : num_threads_(num_threads), api_(Api::createApiForTest()),
+      thread_factory_(api_->threadFactory()) {
+  // This is the same order as InstanceImpl::initialize in source/server/server.cc.
+  thread_dispatchers_.resize(num_threads_);
+  {
+    BlockingBarrier blocking_barrier(num_threads_ + 1);
+    main_thread_ = thread_factory_.createThread(
+        [this, &blocking_barrier]() { mainThreadFn(blocking_barrier); });
+    for (uint32_t i = 0; i < num_threads_; ++i) {
+      threads_.emplace_back(thread_factory_.createThread(
+          [this, i, &blocking_barrier]() { workerThreadFn(i, blocking_barrier); }));
+    }
+  }
+  runOnMainBlocking([this]() {
+    tls_ = std::make_unique<ThreadLocal::InstanceImpl>();
+    tls_->registerThread(*main_dispatcher_, true);
+    for (Event::DispatcherPtr& dispatcher : thread_dispatchers_) {
+      // Worker threads must be registered from the main thread, per assert in registerThread().
+      tls_->registerThread(*dispatcher, false);
+    }
+  });
+}
+
+void RealThreadsTestBase::shutdownThreading() {
+  runOnMainBlocking([this]() {
```

Comment:
It's more complex than desired, I see, because there are some slight differences in shutdown sequence between the two tests that use this.

Maybe you can leave this as is for now but leave a TODO to refactor slightly to avoid repeated code. E.g. maybe you could call a virtual hook to do the extra shutdown call needed in thread_local_store_test.cc but otherwise share the shutdownThreading impl which is almost identical.

I'd be fine having that as a TODO . The right structure is in place for sharing at least.

Reply:
I was also thinking about a virtual override to customize the resource management. But this is intended to be called in dtor, virtual functions shouldn't be called in a destructor.

Reply:
clang-tidy actually enforce that we don't call virtual function from dtor, my previous code result in CI failure.

Reply:
Like I said I'm OK with a TODO to clean this up. This is clearly heading in the right direction.

As indicated above I think this might be better as a helper class rather than an inheritance.

Reply:
Because we access many protected data members of the class directly, if we don't inherit from it, we need to make these members public or friend the class in each class that use it, which doesn't seem right to me.

Reply:
Yeah this is an indication that the refactoring isn't shaped right yet. Let's just drop in a TODO and follow-up to clean it up.